lib/demuxeres/gstdemuxeres.c

```diff
@@ -0,0 +1,721 @@
+      g_free (stream->data.video.profile);
+      g_free (stream->data.video.level);
+      break;
+    default:
+      break;
+  }
+  g_free (stream->stream_id);
+  g_free (stream);
+}
+
+GstDemuxerEStream *
+gst_parse_stream_create (GstDemuxerES * demuxer, GstPad * pad)
+{
+  GstDemuxerEStream *stream;
+  GstPad *funnel_pad;
+  GstDemuxerESPrivate *priv = demuxer->priv;
+  GstCaps *caps = gst_pad_query_caps (pad, NULL);
+
+
+  if (!caps) {
+    GST_ERROR
+        ("Unable to get the caps from pad, unable to create a new stream");
+    return NULL;
+  }
+  GST_INFO ("the stream caps is %" GST_PTR_FORMAT, caps);
+
+  stream = g_new0 (GstDemuxerEStream, 1);
+  stream->demuxer = demuxer;
+  stream->type = gst_parse_stream_get_type_from_pad (pad);
+  stream->stream_id = gst_pad_get_stream_id (GST_PAD_CAST (pad));
+  GstStructure *s = gst_caps_get_structure (caps, 0);
+  switch (stream->type) {
+    case DEMUXER_ES_STREAM_TYPE_VIDEO:
+    {
+      gst_video_info_from_caps (&stream->data.video.info, caps);
+      gst_structure_get_int (s, "bitrate", &stream->data.video.bitrate);
+      stream->data.video.profile =
+          g_strdup (gst_structure_get_string (s, "profile"));
+      stream->data.video.level =
+          g_strdup (gst_structure_get_string (s, "level"));
+      stream->data.video.vcodec = gst_parse_stream_get_vcodec_from_caps (caps);
+      break;
+    }
+    case DEMUXER_ES_STREAM_TYPE_AUDIO:
+    {
+      gst_audio_info_from_caps (&stream->data.audio.info, caps);
+      gst_structure_get_int (s, "bitrate", &stream->data.audio.bitrate);
+      stream->data.audio.acodec = gst_parse_stream_get_acodec_from_caps (caps);
+      break;
+    }
+    default:
+      break;
+  }
+  stream->id = g_list_length (priv->streams);
+
+  funnel_pad = gst_element_request_pad_simple (priv->funnel, "sink_%u");
+
+  if (gst_pad_link (pad, funnel_pad) != GST_PAD_LINK_OK)
+    GST_ERROR ("Unable to plug the pad %p to the capsfilter pad", pad);
+
+  gst_caps_unref (caps);
+
+  gst_object_unref (funnel_pad);
+
+  GST_DEBUG_BIN_TO_DOT_FILE_WITH_TS (GST_BIN (priv->pipeline),
+      GST_DEBUG_GRAPH_SHOW_ALL, "gst-demuxerer.stream_create");
+
+  return stream;
+}
+
+static inline gboolean
+wait_for_demuxer_ready (GstDemuxerES * demuxer)
+{
+  GstDemuxerESPrivate *priv = demuxer->priv;
+
+  g_mutex_lock (&priv->ready_mutex);
+  while (priv->state == DEMUXER_ES_STATE_IDLE) {
+    g_cond_wait (&priv->ready_cond, &priv->ready_mutex);
+  }
+  g_mutex_unlock (&priv->ready_mutex);
+
+  return (priv->state >= DEMUXER_ES_STATE_READY);
+}
+
+static void
+uridecodebin_pad_added_cb (GstElement * uridecodebin, GstPad * pad,
+    GstDemuxerES * demuxer)
+{
+  GstDemuxerEStream *stream = NULL;
+
+  if (!GST_PAD_IS_SRC (pad))
+    return;
+
+  GST_DEBUG ("pad %s:%s", GST_DEBUG_PAD_NAME (pad));
+
+  stream = gst_parse_stream_create (demuxer, pad);
+  if (!stream)
+    return;
+
+  demuxer->priv->streams = g_list_append (demuxer->priv->streams, stream);
```

Comment:
imo, this should be handled in GstStream event

Reply:
See https://github.com/Igalia/GstVkVideoParser/pull/32#discussion_r1065743933